subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache
+
+import org.gradle.configurationcache.initialization.ConfigurationCacheStartParameter
+import org.gradle.internal.hash.HashCode
+import org.gradle.internal.hash.Hashing
+import org.gradle.internal.service.scopes.Scopes
+import org.gradle.internal.service.scopes.ServiceScope
+import org.gradle.util.internal.EncryptionAlgorithm
+import org.gradle.util.internal.EncryptionAlgorithm.EncryptionException
+import org.gradle.util.internal.SupportedEncryptionAlgorithm
+import java.io.InputStream
+import java.io.OutputStream
+import java.security.InvalidKeyException
+import java.util.Base64
+import javax.crypto.CipherInputStream
+import javax.crypto.CipherOutputStream
+import javax.crypto.SecretKey
+import javax.crypto.spec.SecretKeySpec
+
+
+/**
+ * Provides access to configuration parameters to control encryption.
+ */
+internal
+interface EncryptionConfiguration {
+    val isEncrypting: Boolean
+    val encryptionKeyHashCode: HashCode
+    val encryptionAlgorithm: EncryptionAlgorithm
+}
+
+
+enum class EncryptionStrategy(val id: Int, val encrypted: Boolean) {
+    None(0, false),
+    Streams(1, true);
+}
+
+
+/**
+ * A service for encrypting/decrypting streams.
+ */
+internal
+interface EncryptionService : EncryptionConfiguration {
+    fun outputStream(stateType: StateType, output: () -> OutputStream): OutputStream
+    fun inputStream(stateType: StateType, input: () -> InputStream): InputStream
+}
+
+
+@ServiceScope(Scopes.BuildTree::class)
+internal
+class DefaultEncryptionService(startParameter: ConfigurationCacheStartParameter) : EncryptionService {
+
+    private
+    val encryptionStrategy: EncryptionStrategy = if (startParameter.encryptionRequested) EncryptionStrategy.Streams else EncryptionStrategy.None
```

Comment:
❓ This is information is available at the point of service instantiation, isn't it? Why not return a different service implementation at that point?

Reply:
Just dropped the concept of encryption strategy, since there is only one now anyways.